tools/astarte_e2e/lib/astarte_e2e/client.ex

```diff
@@ -0,0 +1,575 @@
+        simple_trigger: %{
+          type: "data_trigger",
+          on: "incoming_data",
+          interface_name: "*",
+          interface_major: 1,
+          match_path: "/*",
+          value_match_operator: "*"
+        }
+      }
+    ]
+
+    with :ok <- install_triggers(device_triggers, transport, state),
+         :ok <- install_triggers(data_triggers, transport, state) do
+      Logger.info("Triggers installed.", tag: "triggers_installed")
+      :ok
+    else
+      {:error, reason} ->
+        Logger.warn("Failed to install triggers with reason: #{inspect(reason)}.",
+          tag: "install_triggers_failed"
+        )
+
+        {:error, reason}
+    end
+  end
+
+  defp install_triggers(triggers, transport, %{callback_state: %{topic: topic}} = _state) do
+    Enum.reduce_while(triggers, :ok, fn trigger, _acc ->
+      case GenSocketClient.push(transport, topic, "watch", trigger) do
+        {:error, reason} ->
+          Logger.warn("Watch failed with reason: #{inspect(reason)}.",
+            tag: "watch_failed"
+          )
+
+          {:halt, {:error, reason}}
+
+        {:ok, _ref} ->
+          Logger.info("Successful watch request.",
+            tag: "watch_success"
+          )
+
+          {:cont, :ok}
+      end
+    end)
+  end
+
+  defp make_topic(realm, device_id) do
+    room_name = Utils.random_string()
+
+    "rooms:#{realm}:#{device_id}_#{room_name}"
+  end
+
+  defp via_tuple(realm, device_id) do
+    {:via, Registry, {Registry.AstarteE2E, {:client, realm, device_id}}}
+  end
+
+  # Callbacks
+
+  def init(opts) do
+    url = Keyword.fetch!(opts, :url)
+    realm = Keyword.fetch!(opts, :realm)
+    jwt = Keyword.fetch!(opts, :jwt)
+    device_id = Keyword.fetch!(opts, :device_id)
+    check_repetitions = Keyword.fetch!(opts, :check_repetitions)
+
+    topic = make_topic(realm, device_id)
+
+    callback_state = %{
+      device_id: device_id,
+      topic: topic
+    }
+
+    query_params = [realm: realm, token: jwt]
+
+    state = %{
+      callback_state: callback_state,
+      pending_requests: %{},
+      pending_messages: %{},
+      connection_attempts: @connection_attempts,
+      check_repetitions: check_repetitions,
+      waiting_for_connection: %{},
+      timeouts_to_crash: Config.client_max_timeouts!(),
+      connected: false
+    }
+
+    {:connect, url, query_params, state}
+  end
+
+  def handle_connected(transport, state) do
+    Logger.info("Connected.", tag: "client_connected")
+    state = %{state | connection_attempts: @connection_attempts, connected: true}
+
+    waiting_for_connection = state.waiting_for_connection
+
+    new_waiting =
+      if Map.has_key?(waiting_for_connection, self()) do
+        {tref, new_waiting} = Map.pop!(waiting_for_connection, self())
+        :ok = Process.cancel_timer(tref, async: false, info: false)
+        new_waiting
+      else
+        waiting_for_connection
+      end
+
+    new_state = %{state | waiting_for_connection: new_waiting}
```

Comment:
Update the state in one shot (i.e. update here also `connection_attempts` and `connected` instead of doing it at the beginning). This helps seeing all that gets updated in a single glance.